pkg/reboot/calculator.go

```diff
@@ -106,14 +115,79 @@ func (s *safeTimeCalculator) calcMinTimeAssumeRebooted() error {
 	s.log.Info("calculated minTimeToAssumeNodeRebooted is:", "minTimeToAssumeNodeRebooted", minTime)
 	s.minTimeToAssumeNodeRebooted = minTime
 
-	if s.timeToAssumeNodeRebooted < minTime {
-		err := fmt.Errorf("snr agent can't start: the requested value for SafeTimeToAssumeNodeRebootedSeconds is too low")
-		s.log.Error(err, err.Error(), "requested SafeTimeToAssumeNodeRebootedSeconds", s.timeToAssumeNodeRebooted, "minimal calculated value for SafeTimeToAssumeNodeRebootedSeconds", minTime)
+	//update related logic of min time on configuration if necessary
+	if err := s.manageSafeRebootTimeInConfiguration(minTime); err != nil {
 		return err
 	}
+
 	return nil
 }
 
+// manageSafeRebootTimeInConfiguration does two things:
+//  1. It sets Status.MinSafeTimeToAssumeNodeRebootedSeconds in case it's changed by latest calculation.
+//  2. It Adds/removes config.Status.SpecSafeTimeOverriddenWarning if necessary.
+func (s *safeTimeCalculator) manageSafeRebootTimeInConfiguration(minTime time.Duration) error {
+	minTimeSec := int(minTime.Seconds())
+	config, err := s.getConfig()
+	if err != nil {
+		return err
+	}
+	orgConfig := config.DeepCopy()
+	prevMinRebootTimeSec := config.Status.MinSafeTimeToAssumeNodeRebootedSeconds
+	if prevMinRebootTimeSec != minTimeSec {
+		config.Status.MinSafeTimeToAssumeNodeRebootedSeconds = minTimeSec
+	}
+	//Manage warning
+	config.Status.SpecSafeTimeOverriddenWarning = ""
+	if config.Spec.SafeTimeToAssumeNodeRebootedSeconds > 0 && minTimeSec > config.Spec.SafeTimeToAssumeNodeRebootedSeconds {
+		config.Status.SpecSafeTimeOverriddenWarning = v1alpha1.SafeTimeToAssumeNodeRebootedSecondsWarning
+		s.log.Info("SafeTimeToAssumeNodeRebootedSeconds is overridden by calculated value because it's invalid, calculated value stored in Status.MinSafeTimeToAssumeNodeRebootedSeconds would be used instead")
+		events.WarningEvent(s.recorder, config, "SafeTimeToAssumeNodeRebootedSecondsInvalid", "SafeTimeToAssumeNodeRebootedSeconds is overridden by calculated value")
+	}
+
+	if !reflect.DeepEqual(config, orgConfig) {
+		if err := s.k8sClient.Status().Patch(context.Background(), config, client.MergeFrom(orgConfig)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *safeTimeCalculator) getMinSafeTimeSecFromConfig() (time.Duration, error) {
+	config, err := s.getConfig()
+	if err != nil {
+		return 0, err
+	}
+
+	minTimeSec := config.Status.MinSafeTimeToAssumeNodeRebootedSeconds
+	if minTimeSec > 0 {
+		return time.Duration(minTimeSec) * time.Second, nil
+	}
+
+	err = errors.New("failed getting MinSafeTimeToAssumeNodeRebootedSeconds, value isn't initialized")
+	s.log.Error(err, "MinSafeTimeToAssumeNodeRebootedSeconds should not be empty")
+	return 0, err
+}
+
+func (s *safeTimeCalculator) getConfig() (*v1alpha1.SelfNodeRemediationConfig, error) {
+	confList := &v1alpha1.SelfNodeRemediationConfigList{}
+	if err := s.k8sClient.List(context.Background(), confList); err != nil {
```

Comment:
please use an existing context...

Reply:
In general I agree it's better practice to use the existing context over creating one.

In this specific use case it would mean adding that context to at least another 8 methods that does not use it (expect for passing it), which is a downside IMO as far code readability/complexity goes.

Since in addition, in this specific use case we don't utilize any capabilities which needs the original context, IMO it's a better trade-off to keep the code as is.

Let me know if that's something you feel strongly about.

Reply:
>  we don't utilize any capabilities which needs the original context

We do, in the List method. When the original context is cancelled because e.g. the pod is shutting down, the List method is cancelled as well, instead of waiting for it to finish, which potentially can take some while in case of problems.

>  adding that context to at least another 8 methods that does not use it

looking at the controller code, many methods don't have a context as arg yet, because there many places which use `TODO` or `Background` context... 😞